### PR TITLE
Hash Join: improve offset calculation

### PR DESCRIPTION
**Pull request by Bouncner**
As found by @mrks, the offset calculation of the hash join can be quite slow for very large inputs (e.g., TPC-H with a scale factor of 1000).

This PR changes the previous code that was rather randomly aggregating values and thus created a lot of cache misses. The new code iterates the input data twice, but fully sequential.

The change is not measurable for small joins.
For a SF=100 join and Q21, the offset part of the largest join went from 1.75s and 1.57s to now 0.29s and 0.28s. For larger inputs, the change should be more drastic.

### CODE REVIEW

File: src/lib/operators/join_hash/join_hash_steps.hpp [hunk 1]
```diff
@@ -338,34 +338,77 @@ RadixContainer<T> partition_radix_parallel(const RadixContainer<T>& radix_contai
...
-  // use histograms to calculate partition offsets
-  size_t offset = 0;
-  std::vector<std::vector<size_t>> output_offsets_by_chunk(chunk_offsets.size(), std::vector<size_t>(num_partitions));
-  for (size_t partition_id = 0; partition_id < num_partitions; ++partition_id) {
-    for (ChunkID chunk_id{0}; chunk_id < chunk_offsets.size(); ++chunk_id) {
-      output_offsets_by_chunk[chunk_id][partition_id] = offset;
-      offset += histograms[chunk_id][partition_id];
+  /**
+   * The following steps create the offsets that allow each concurrent job to write lock-less into the newly
+   * created RadixContainer. The input RadixContainer was a simple list materialized values. The outcome
```
Comment (Bensk1) [inline src/lib/operators/join_hash/join_hash_steps.hpp]: simple list **of** materialized values?

File: src/lib/operators/join_hash/join_hash_steps.hpp [hunk 2]
```diff
@@ -338,34 +338,77 @@ RadixContainer<T> partition_radix_parallel(const RadixContainer<T>& radix_contai
...
+   * container is consectitive vector, sorted by radix clusters (which are defined by the offsets).
```
Comment (Bensk1) [inline src/lib/operators/join_hash/join_hash_steps.hpp]: consectitive

File: src/lib/operators/join_hash/join_hash_steps.hpp [hunk 3]
```diff
@@ -338,34 +338,77 @@ RadixContainer<T> partition_radix_parallel(const RadixContainer<T>& radix_contai
...
+   * Input:
+   *  - `histograms` stores for each input chunk the histogram that counts the number of value per radix cluster.
```
Comment (Bensk1) [inline src/lib/operators/join_hash/join_hash_steps.hpp]: values

File: src/lib/operators/join_hash/join_hash_steps.hpp [hunk 4]
```diff
@@ -338,34 +338,77 @@ RadixContainer<T> partition_radix_parallel(const RadixContainer<T>& radix_contai
...
+   *  - `chunk_offsets` store the offsets -- denoting the number of elements of each chunk -- in the continuous
```
Comment (Bensk1) [inline src/lib/operators/join_hash/join_hash_steps.hpp]: stores

File: src/lib/operators/join_hash/join_hash_steps.hpp [hunk 5]
```diff
@@ -338,34 +338,77 @@ RadixContainer<T> partition_radix_parallel(const RadixContainer<T>& radix_contai
...
+   *     vector of materialized values.
+   *
+   * Simple example:
+   * - histograms for chunks 0 and 1 [4 0 3] & [5 2 7] (i.e., first radix cluster has 4 + 5 values)
+   * - first step: create offsets that denote write offsets per radix cluster and collect lengths
+   *   - result: offsets vectors are [0 0 0] [4 0 3] and lengths are [9 2 10]
+   * - second step: create prefix sum vector of [9 2 10] >> [9 11 10]
+   * - third step: adapt offset vectors to create the offsets that allows writing
+   *   all thread in parallel into a _consecutive_ vector
```
Comment (Bensk1) [inline src/lib/operators/join_hash/join_hash_steps.hpp]: threads

File: src/lib/operators/join_hash/join_hash_steps.hpp [hunk 6]
```diff
@@ -338,34 +338,77 @@ RadixContainer<T> partition_radix_parallel(const RadixContainer<T>& radix_contai
...
+   *  - `histograms` stores for each input chunk the histogram that counts the number of value per radix cluster.
+   *  - `chunk_offsets` store the offsets -- denoting the number of elements of each chunk -- in the continuous
+   *     vector of materialized values.
+   *
+   * Simple example:
+   * - histograms for chunks 0 and 1 [4 0 3] & [5 2 7] (i.e., first radix cluster has 4 + 5 values)
+   * - first step: create offsets that denote write offsets per radix cluster and collect lengths
+   *   - result: offsets vectors are [0 0 0] [4 0 3] and lengths are [9 2 10]
+   * - second step: create prefix sum vector of [9 2 10] >> [9 11 10]
+   * - third step: adapt offset vectors to create the offsets that allows writing
```
Comment (Bensk1) [inline src/lib/operators/join_hash/join_hash_steps.hpp]: allow?

File: src/lib/operators/join_hash/join_hash_steps.hpp [hunk 7]
```diff
@@ -338,34 +338,77 @@ RadixContainer<T> partition_radix_parallel(const RadixContainer<T>& radix_contai
...
+   *   all thread in parallel into a _consecutive_ vector
+   *   - result: [0 9 11] [4 9 14]
+   * - note: the third step would be superflous, when each radix cluster is written to separate vector
```
Comment (Bensk1) [inline src/lib/operators/join_hash/join_hash_steps.hpp]: superfluous

Comment (Bensk1) [reply src/lib/operators/join_hash/join_hash_steps.hpp]: **a** separate vector

**Comment by Bouncner** [comment]
Ouch. I need to integrate some serious grammar checker in Sublime. :-/

Will fix them, thanks.

File: src/lib/operators/join_hash/join_hash_steps.hpp [hunk 8]
```diff
@@ -338,34 +338,77 @@ RadixContainer<T> partition_radix_parallel(const RadixContainer<T>& radix_contai
...
-  // use histograms to calculate partition offsets
-  size_t offset = 0;
-  std::vector<std::vector<size_t>> output_offsets_by_chunk(chunk_offsets.size(), std::vector<size_t>(num_partitions));
-  for (size_t partition_id = 0; partition_id < num_partitions; ++partition_id) {
-    for (ChunkID chunk_id{0}; chunk_id < chunk_offsets.size(); ++chunk_id) {
-      output_offsets_by_chunk[chunk_id][partition_id] = offset;
-      offset += histograms[chunk_id][partition_id];
+  /**
+   * The following steps create the offsets that allow each concurrent job to write lock-less into the newly
+   * created RadixContainer. The input RadixContainer was a simple list of materialized values. The outcome
```
Comment (mrks) [inline src/lib/operators/join_hash/join_hash_steps.hpp]: why "was"?

File: src/lib/operators/join_hash/join_hash_steps.hpp [hunk 9]
```diff
@@ -338,34 +338,77 @@ RadixContainer<T> partition_radix_parallel(const RadixContainer<T>& radix_contai
...
+   * container is consecutive vector, sorted by radix clusters (which are defined by the offsets).
```
Comment (mrks) [inline src/lib/operators/join_hash/join_hash_steps.hpp]: is **a** consecutive vector

File: src/lib/operators/join_hash/join_hash_steps.hpp [hunk 10]
```diff
@@ -338,34 +338,77 @@ RadixContainer<T> partition_radix_parallel(const RadixContainer<T>& radix_contai
...
+   *  - `chunk_offsets` stores the offsets -- denoting the number of elements of each chunk -- in the continuous
+   *     vector of materialized values.
+   *
+   * Simple example:
+   * - histograms for chunks 0 and 1 [4 0 3] & [5 2 7] (i.e., first radix cluster has 4 + 5 values)
+   * - first step: create offsets that denote write offsets per radix cluster and collect lengths
+   *   - result: offsets vectors are [0 0 0] [4 0 3] and lengths are [9 2 10]
+   * - second step: create prefix sum vector of [9 2 10] >> [9 11 10]
+   * - third step: adapt offset vectors to create the offsets that allow writing
+   *   all threads in parallel into a _consecutive_ vector
```
Comment (mrks) [inline src/lib/operators/join_hash/join_hash_steps.hpp]: Vectors are always contiguous

File: src/lib/operators/join_hash/join_hash_steps.hpp [hunk 11]
```diff
@@ -338,34 +338,77 @@ RadixContainer<T> partition_radix_parallel(const RadixContainer<T>& radix_contai
...
+   *   - result: [0 9 11] [4 9 14]
+   * - note: the third step would be superfluous, when each radix cluster is written to a separate vector
```
Comment (mrks) [inline src/lib/operators/join_hash/join_hash_steps.hpp]: no comma

File: src/lib/operators/join_hash/join_hash_steps.hpp [hunk 12]
```diff
@@ -338,34 +338,77 @@ RadixContainer<T> partition_radix_parallel(const RadixContainer<T>& radix_contai
...
+   * Input:
+   *  - `histograms` stores for each input chunk a histogram that counts the number of values per radix cluster.
+   *  - `chunk_offsets` stores the offsets -- denoting the number of elements of each chunk -- in the continuous
+   *     vector of materialized values.
+   *
+   * Simple example:
+   * - histograms for chunks 0 and 1 [4 0 3] & [5 2 7] (i.e., first radix cluster has 4 + 5 values)
+   * - first step: create offsets that denote write offsets per radix cluster and collect lengths
+   *   - result: offsets vectors are [0 0 0] [4 0 3] and lengths are [9 2 10]
+   * - second step: create prefix sum vector of [9 2 10] >> [9 11 10]
```
Comment (mrks) [inline src/lib/operators/join_hash/join_hash_steps.hpp]: 21

File: src/lib/operators/join_hash/join_hash_steps.hpp [hunk 13]
```diff
@@ -338,34 +338,77 @@ RadixContainer<T> partition_radix_parallel(const RadixContainer<T>& radix_contai
...
+   */
+  std::vector<size_t> output_offsets_by_chunk;
+  output_offsets_by_chunk.reserve(chunk_offsets.size() * num_partitions);
+
+  // Offset vector creation: first step
+  auto prefix_sums = std::vector<size_t>(num_partitions);  // stores the prefix sums
+  for (auto chunk_id = ChunkID{0}; chunk_id < chunk_offsets.size(); ++chunk_id) {
+    for (auto radix_cluster_id = size_t{0}; radix_cluster_id < num_partitions; ++radix_cluster_id) {
+      const auto radix_cluster_size = histograms[chunk_id][radix_cluster_id];
+      output_offsets_by_chunk.push_back(prefix_sums[radix_cluster_id]);
```
Comment (mrks) [inline src/lib/operators/join_hash/join_hash_steps.hpp]: Do we need push_back or can we use zero-initialization and indexes?

File: src/lib/operators/join_hash/join_hash_steps.hpp [hunk 14]
```diff
@@ -338,34 +338,77 @@ RadixContainer<T> partition_radix_parallel(const RadixContainer<T>& radix_contai
...
+   * - second step: create prefix sum vector of [9 2 10] >> [9 11 10]
+   * - third step: adapt offset vectors to create the offsets that allow writing
+   *   all threads in parallel into a _consecutive_ vector
+   *   - result: [0 9 11] [4 9 14]
+   * - note: the third step would be superfluous, when each radix cluster is written to a separate vector
+   */
+  std::vector<size_t> output_offsets_by_chunk;
+  output_offsets_by_chunk.reserve(chunk_offsets.size() * num_partitions);
+
+  // Offset vector creation: first step
```
Comment (mrks) [inline src/lib/operators/join_hash/join_hash_steps.hpp]: Explain why we have multiple steps?

File: src/lib/operators/join_hash/join_hash_steps.hpp (hunk 10 above)
Comment (Bouncner) [reply src/lib/operators/join_hash/join_hash_steps.hpp]: Changed to "single".

File: src/lib/operators/join_hash/join_hash_steps.hpp (hunk 14 above)
Comment (Bouncner) [reply src/lib/operators/join_hash/join_hash_steps.hpp]: Added the following comment in the comment block:
```
* The process of creating the offset information consists of three steps. A previous commit used a single
   * loop and created multiple vectors. This approach has shown to be inefficient for very large inputs.
   * Consequently, we now use a large single vector and operate fully sequentially. This come at the cost of
   * iterating twice and introduce additional steps. However, the current approach should be faster.
```

Does that help?

File: src/lib/operators/join_hash/join_hash_steps.hpp (hunk 13 above)
Comment (Bouncner) [reply src/lib/operators/join_hash/join_hash_steps.hpp]: Done.

**Comment by mrks** [comment]
As this takes forever, I didn't want to re-run the vanilla SF 1000 version to measure the cost of this step, so here are numbers from end to end:

SF 1000, rapa, all cores, single client
old: `Executed 1 times in 2039.71 seconds (0.000490266 iter/s)`
new: `Executed 1 times in 1885.42 seconds (0.000530387 iter/s)`

Nice :)

**Pull request merged**